### PR TITLE
Add Pending Activity check

### PR DESCRIPTION
**Pull request by DeepanshuA**
# Description

This PR addresses issue occurring while re-scheduling a workflow with same instance id, while previous run was terminated but still an activity left un-finished.

I have tried to enlist the detailed description of issue and thought process behind fix in following comment: https://github.com/dapr/dapr/issues/6454#issuecomment-1670041771

Another workflow with same instance Id is allowed to Run.
That's why the concept of Generation in there. And also, in https://github.com/dapr/dapr/pull/6455#issue-1745212239 , @cgillum specifically talks about it, so that means that it is Allowed.

To keep behavior consistent, I strongly feel that thus any Workflow terminated once, should be allowed to be scheduled with same id again.
It is just that the definition of Complete needs to be defined correctly i.e. Workflow with same ID should not be allowed just because previous Workflow run was Terminated. It should be allowed only if No Activity is remaining to be returned back as Completed or Failed, in addition to Workflow Completed.

So, let's analyze following 2 cases:

Case 1:
Scheduled a workflow W1 with activities A1, A2, A3 and A4. Let's Assume, workflow id is ID1.
In A2, put a sleep of 30 seconds.
After 10 seconds of scheduling workflow, Terminate it.
And, schedule this workflow again with same id ID1.

Here, re-scheduling of this workflow should NOT be allowed as A2 has not yet returned, even though Workflow itself has been Terminated.

Case 2:
Scheduled a workflow W1 with activities A1, A2, A3 and A4. Let's Assume, workflow id is ID1.
In A2, put a sleep of 30 seconds.
After 10 seconds of scheduling workflow, Terminate it.
Wait for 25 more seconds. By this time, A2 has also Completed. A3 and A4 will not be called.
At this time, schedule this workflow again with same id ID1.

Here, re-scheduling of this workflow SHOULD be allowed as A2 has returned, and Workflow itself has been truly Terminated and Completed.

## Issue reference

Please reference the issue this PR will close: #6454 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/runtime/wfengine/workflow.go [hunk 1]
```diff
@@ -175,12 +176,17 @@ func (wf *workflowActor) createWorkflowInstance(ctx context.Context, actorID str
 		return errors.New("invalid execution start event")
 	}
 
-	// We block (re)creation of existing workflows unless they are in a completed state.
+	// We block (re)creation of existing workflows unless they are in a completed state
+	// Or if they still have any pending activity result awaited.
 	if !created {
```
Comment (mukundansundar) [inline pkg/runtime/wfengine/workflow.go:181]: change flow as such:
```go
  if created {
     return error
  }
  // unnest the rest 
  runtimeState := getRuntimeState(actorID, state)
  if !runtimeState.IsCompleted() {
     return error
  }
  if wf.activityResultAwaited {
     return fmt.Errorf("a terminated workflow with ID '%s' is already awaiting an activity result", actorID)
  }
  // no else needed
  wfLogger.Infof("%s: workflow was previously completed and is being recreated", actorID)
  state.Reset()
```

File: pkg/runtime/wfengine/workflow.go [hunk 2]
```diff
@@ -440,6 +449,8 @@ func (wf *workflowActor) runWorkflow(ctx context.Context, actorID string, remind
 				WithContentType(invokev1.OctetStreamContentType)
 			defer req.Close()
 
+			wf.activityResultAwaited = true
```
Comment (mukundansundar) [inline pkg/runtime/wfengine/workflow.go:452]: will this state be accessed by multiple goroutines? Does this need to be an atomic boolean?

File: pkg/runtime/wfengine/workflow.go (hunk 1 above)
Comment (DeepanshuA) [reply pkg/runtime/wfengine/workflow.go:182]: Could not take the first check from suggestion:
```
  if created {
     return error
  }
```
I have taken the later part of suggestion, that sounds good.

**Review by mukundansundar — APPROVED** [review]
lgtm

**Pull request merged**